### PR TITLE
Add Kerberos support to Elasticsearch output

### PR DESCRIPTION
**Pull request by kvch**
## What does this PR do?

This PR adds support for Kerberos authentication to Elasticsearch output.

### Configuration

Users can authenticate using either passwords or keytabs.

The option `service_name` is not exposed as in case of ES it has be `HTTP`. Thus, the [SPN](https://web.mit.edu/kerberos/krb5-1.5/krb5-1.5.4/doc/krb5-user/What-is-a-Kerberos-Principal_003f.html) of the output is always `HTTP/{output.elasticsearch.host}@{output.elasticsearch.kerberos.realm}`.

```yaml
# Authentication type to use with Kerberos. Available options: keytab, password.
#kerberos.auth_type: password

# Path to the keytab file. It is used when auth_type is set to keytab.
#kerberos.keytab: /etc/elastic.keytab

# Path to the Kerberos configuration.
#kerberos.config_path: /etc/krb5.conf

# Name of the Kerberos user.
#kerberos.username: elastic

# Password of the Kerberos user. It is used when auth_type is set to password.
#kerberos.password: changeme

# Kerberos realm.
#kerberos.realm: ELASTIC
```

### Architecture

Kerberos authentication is implemented by using the `spnego.Client` from `gopkg.in/jcmturner/gokrb5/v7.5`. It is an HTTP client which is able to authenticate to a Kerberos [KDC](https://en.wikipedia.org/wiki/Key_distribution_center) and get the required session key to authenticate to Elasticsearch.

In order to use the client when communicating with ES, I introduced a new interface named `esHTTPClient` with the following functions: 

```golang
type esHTTPClient interface {
	Do(req *http.Request) (resp *http.Response, err error)
	CloseIdleConnections()
}
```

So the newly created `kerberos.Client` and `http.Client` can be used in `Connection`.

The only drawback of this solution is that it is not compatible with `github.com/elastic/go-elasticsearch`. The main issue is the constructor of that package only lets us configure `http.Transport`. However, `http.Transport` should be used for authentication according to Golang documentation. For the record, it is possible to implement Kerberos authentication in a special `RoundTripper`, but I am not a huge fan of that approach as the Golang docs advise against it.

Reference:
- [SPNEGO client](https://pkg.go.dev/github.com/jcmturner/gokrb5@v7.5.0+incompatible/spnego?tab=doc#Client)
- [http.Transport](https://golang.org/pkg/net/http/#Transport)

### Testing

I am planning to add tests in a follow-up PR. I wanted to raise this first, so we can agree on the selected architecture.

## Why is it important?

Users now can authenticate to Kerberos-aware Elasticsearch instances.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

File: libbeat/_meta/config.reference.yml.tmpl [hunk 1]
```diff
@@ -1081,6 +1081,23 @@ setup.kibana:
...
+  #kerberos.config_path: /etc/krb5.conf
+
+  # Name of the Kerberos user.
+  #kerberos.username: elastic
+
+  # Password of the Kerberos user. It is used when auth_type is set to password.
+  #kerberos.password: changeme
+
+  # Kerberos realm.
+  #kerberos.realm: ELASTIC
```
Comment (ycombinator) [inline libbeat/_meta/config.reference.yml.tmpl:1100]: These are all under `setup.kibana`. Is that correct or did you mean to put them under `output.elasticsearch` (or both)?

Comment (kvch) [reply libbeat/_meta/config.reference.yml.tmpl:1100]: :facepalm: :facepalm: :facepalm:

Comment (kvch) [reply libbeat/_meta/config.reference.yml.tmpl:1100]: Fixed. https://images-cdn.9gag.com/photo/6324230_700b.jpg :D :D :D

File: filebeat/docs/fields.asciidoc [hunk 2]
```diff
@@ -28240,6 +28240,15 @@ type: keyword
...
 
+type: keyword
+
+--
+
+*`panw.panos.action`*::
++
+--
+Action taken for the session.
+
```
Comment (ycombinator) [inline filebeat/docs/fields.asciidoc:28251]: Should this change be in this PR? I think I saw a separate PR recently to fix this: https://github.com/elastic/beats/pull/17947/files.

Comment (kvch) [reply filebeat/docs/fields.asciidoc]: Great. I did not know about the fix. Thank you.

**Comment by ycombinator** [comment]
> The only drawback of this solution is that it is not compatible with `github.com/elastic/go-elasticsearch`. The main issue is the constructor of that package only lets us configure `http.Transport`. However, `http.Transport` should be used for authentication according to Golang documentation. For the record, it is possible to implement Kerberos authentication in a special `RoundTripper`, but I am not a huge fan of that approach as the Golang docs advise against it.

WDYT about starting a discussion issue about this in https://github.com/elastic/go-elasticsearch?

File: libbeat/common/transport/kerberos/config.go [hunk 3]
```diff
@@ -59,19 +64,21 @@ func (t *AuthType) Unpack(value string) error {
 }
 
 func (c *Config) Validate() error {
-	if c.AuthType == AUTH_PASSWORD {
+	switch c.AuthType {
+	case AUTH_PASSWORD:
```
Comment (urso) [inline libbeat/common/transport/kerberos/config.go:68]: nits: Do not use C like screaming identifiers for consts. Plus add the actual type to the const if we already have an enum type. Do we need to export the consts or not?

I understand these have not been introduced in this PR, but would be nice to cleanup existing code a little when we touch it :)

**Comment by urso** [comment]
How about adding an `enabled` setting to the kerberos configuration namespace? We have `enabled` for almost everything else, even `ssl` settings have an `enabled` setting.

File: libbeat/common/transport/kerberos/config.go (hunk 3 above)
Comment (ycombinator) [reply libbeat/common/transport/kerberos/config.go]: TIL the all-caps identifiers are also known as screaming identifiers! 😄

**Comment by urso** [comment]
LGTM. Can you please add a changelog?

File: libbeat/docs/shared-kerberos-config.asciidoc [hunk 4]
```diff
@@ -0,0 +1,85 @@
+[[configuration-kerberos]]
```
Comment (bmorelli25) [inline libbeat/docs/shared-kerberos-config.asciidoc:1]: It looks like you haven't included this file anywhere. It's not being built, which is why the link [here](https://github.com/elastic/beats/pull/17927/files#diff-1eea36cd1513c862a97a56d38522dd58R684) fails.

Comment (kvch) [reply libbeat/docs/shared-kerberos-config.asciidoc:1]: Thank you for your help!

**Comment by kvch** [comment]
Failing tests are unrelated.

**Pull request merged**